### PR TITLE
feat: add grand exchange trade notifier

### LINKED ISSUE #272: Grand Exchange Bought/Sold Notifications
Would be nice if you could have a bought/sold webhook for grand exchange transactions.

### PR DESCRIPTION
**Pull request by iProdigy**
Closes #272 (Related: https://github.com/OkayestDev/OSRS-Flipper/issues/18)

image

Current rules:
* Transaction must exceed min value
* Cancellation is not notified unless enabled in config
* Partial progress is not notified unless spacing minutes have passed

Required feedback:
* Are the above rules reasonable
* Do the rich embed fields need any changes
* Does more real-world testing reveal any edge cases

Sample:
```json
{
  "type": "GRAND_EXCHANGE",
  "playerName": "dankgg",
  "accountType": "NORMAL",
  "extra": {
    "slot": 2,
    "status": "BOUGHT",
    "item": {
      "id": 314,
      "quantity": 2,
      "priceEach": 3,
      "name": "Feather"
    },
    "marketPrice": 2,
    "targetQuantity": 2
  },
  "embeds": [
    {
      "title": "Grand Exchange",
      "description": "dankgg bought 2 x [Feather](https://oldschool.runescape.wiki/w/Special:Search?search=Feather) (6) on the GE",
      "author": {
        "name": "dankgg",
        "url": "https://secure.runescape.com/m=hiscore_oldschool/hiscorepersonal?user1=dankgg"
      },
      "color": 15990936,
      "image": {
        "url": "attachment://grandExchange.png"
      },
      "thumbnail": {
        "url": "https://oldschool.runescape.wiki/images/Grand_Exchange_icon.png"
      },
      "fields": [
        {
          "name": "Status",
          "value": "```\nCompleted\n```",
          "inline": true
        },
        {
          "name": "Market Unit Price",
          "value": "```ldif\n2 gp\n```",
          "inline": true
        }
      ],
      "footer": {
        "text": "Powered by Dink",
        "icon_url": "https://github.com/pajlads/DinkPlugin/raw/master/icon.png"
      },
      "timestamp": "2023-07-06T06:21:28.672174400Z"
    }
  ]
}
```

### CODE REVIEW

**Comment by Felanbird** [comment]
Is `slot` the spot in the GE in which the item was sold from? I see that the feathers in your example are `"slot": 1,` but are in what I expect to be slot 2.

**Comment by iProdigy** [comment]
It's zero indexed; could prolly document that in readme

File: src/main/java/dinkplugin/message/NotificationType.java [hunk 1]
```diff
@@ -22,7 +22,8 @@ public enum NotificationType {
     ACHIEVEMENT_DIARY("Achievement Diary", "achievementDiaryImage.png", WIKI_IMG_BASE_URL + "Achievement_Diaries_icon.png"),
     BARBARIAN_ASSAULT_GAMBLE("Barbarian Assault Gamble", "baGambleImage.png", WIKI_IMG_BASE_URL + "Barbarian_Assault_logo.jpg"),
     PLAYER_KILL("Player Kill", "playerKillImage.png", WIKI_IMG_BASE_URL + "Skull_(status)_icon.png"),
-    GROUP_STORAGE("Group Shared Storage", "groupStorage.png", WIKI_IMG_BASE_URL + "Coins_10000.png");
+    GROUP_STORAGE("Group Shared Storage", "groupStorage.png", WIKI_IMG_BASE_URL + "Coins_10000.png"),
+    GRAND_EXCHANGE("Grand Exchange", "grandExchange.png", WIKI_IMG_BASE_URL + "Grand_Exchange_icon.png");
```
Comment (iProdigy) [inline src/main/java/dinkplugin/message/NotificationType.java:26]: This icon is rather small, would be nice to find a slightly larger version. https://oldschool.runescape.wiki/w/File:Grand_Exchange_logo.png seems too big

**Comment by iProdigy** [comment]
Known shortcoming: the notifier only triggers on GE events that occur after login has completed (i.e., if a transaction completes while offline, when the player logs in, *no* notification will be fired). This is to avoid duplicate notifications on subsequent logins without collecting the offer proceeds. In a future PR, we can use the saved config data from the GE plugin to check whether offer completions are new or old.

**Review by Felanbird — APPROVED** [review]
Functionality wise I believe this works as expected.

I only have minimal thoughts on multiples of items selling for different values, but in _most_ cases it should not matter if the `priceEach` is slightly wrong, as long as the totals are correct, which I believe in the cases I've seen, it is.

File: src/main/java/dinkplugin/message/NotificationType.java (hunk 1 above)
Comment (pajlada) [reply src/main/java/dinkplugin/message/NotificationType.java:26]: I think using the small icon is the way to go for now, the logo doesn't scale well

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,7 @@
 ## Unreleased
 
+- Major: Add notifier for buying and selling items on the Grand Exchange. (#275)
```
Comment (pajlada) [inline CHANGELOG.md:3]:
```suggestion
- Major: Add notifier for when items are bought or sold on the Grand Exchange. (#275)
```

**Review by pajlada — APPROVED** [review]
Sales & purchases work as expected for me
I think the discord embed looks good, and I don't think the icon looks too small

Cancellations only fire if it's for a partially completed offer, that's kind of written in the config description of "Include Cancelled" but I don't think it's 100% clear.
I think there should either be a line in the description saying: "Cancellation events will only fire if the offer has been partially completed", or you should be able to set "Min Value" to 0, in case you want cancellations of unfulfilled offers to fire too

**Review by pajlada — APPROVED** [review]
